memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressAsyncStorageHandler.java

```diff
@@ -0,0 +1,648 @@
+    executionTimer.schedule(() -> {
+      final TimeoutException ex = new TimeoutException(
+          "Timeout after " + duration.toMillis() + " milliseconds");
+      return promise.completeExceptionally(ex);
+    }, duration.toMillis(), TimeUnit.MILLISECONDS);
+    CompletableFuture<UploadResult> anyUploadResultFuture = CompletableFuture
+        .anyOf(tasks.toArray(new CompletableFuture[0])).thenApply(o -> (UploadResult) o);
+    return anyUploadResultFuture.applyToEither(promise, Function.identity());
+  }
+
+  public void closeWriter() {
+    notificationSink.close();
+  }
+
+  protected KafkaNotificationSink getNotificationSink() {
+    return notificationSink;
+  }
+
+  @Override
+  public String getReadUrl() {
+    return notificationSink.getReadUrl();
+  }
+
+  @Override
+  public Logger getLogger() {
+    return logger;
+  }
+  
+  @Override
+  public InputStream fetchBatchStreamForNotification(JsonObject nextNotificationToProcess) throws IOException {
+    AwsS3V4Signer signer = AwsS3V4Signer.create();
+
+    String currentBucket = nextNotificationToProcess.get(BUCKET).getAsString();
+    String currentKey = nextNotificationToProcess.get(KEY).getAsString();
+    int currentObjectSize = nextNotificationToProcess.get(SIZE).getAsInt();
+    getLogger().fine("Updating bucket and key: " + currentBucket + "/" + currentKey + " {"
+        + nextNotificationToProcess.get(MemqLogMessage.INTERNAL_FIELD_NOTIFICATION_PARTITION_ID)
+            .getAsNumber()
+        + ", " + nextNotificationToProcess
+            .get(MemqLogMessage.INTERNAL_FIELD_NOTIFICATION_PARTITION_OFFSET).getAsNumber()
+        + "}");
+    getLogger().finest("Object size: " + currentObjectSize);
+    
+    SessionTokenManager instance = SessionTokenManager.getInstance();
+    SessionCreds credentials;
+    try {
+      credentials = instance.getCredentials(currentBucket);
+    } catch (InterruptedException e) {
+      throw new IOException(e);
+    }
+    
+    
+    SdkHttpFullRequest req = SdkHttpFullRequest.builder().method(SdkHttpMethod.GET)
+        .appendHeader("x-amz-s3session-token", credentials.token)
+        .uri(URI.create(
+            baseConnStr + currentKey))
+        .build();
+    final SdkHttpFullRequest req1 = signer.sign(req,
+        AwsS3V4SignerParams.builder().awsCredentials(new AwsCredentials() {
+
+          @Override
+          public String secretAccessKey() {
+            return credentials.secret;
+          }
+
+          @Override
+          public String accessKeyId() {
+            return credentials.key;
+          }
+        }).signingName("s3express").signingRegion(region).build());
+    
+    long fetchStartTime = System.currentTimeMillis();
+    try {
+      return httpClient.tryObjectGet(req1);
+    } finally {
+      long fetchTime = System.currentTimeMillis() - fetchStartTime;
+      getLogger().fine("Fetch Time:" + fetchTime);
+      registry.histogram(OBJECT_FETCH_LATENCY_MS_HISTOGRAM_KEY).update(fetchTime);
```

Comment:
Registry can be null and cause NPE here. There's no logic in intilization to make sure the registry is created. 

We can move this into try part or add extra check to prevent NPE.